### DISCUSSION TITLE
关于协程锁的疑惑

### DISCUSSION THREAD

**Discussion by shapled** [idealvin/coost#217] [Q&A] [answered]
我看 Mutex 的实现是用的 pthread_mutex_lock，这个不会锁了 scheduler 吗？还是有更高层次上的处理？

**Comment by idealvin** [comment]
那只是轻粒度的，主要逻辑不会被锁住

**Comment by shapled** [comment]
不是粒度的问题，持有同一把锁的两个协程在一个线程内调度，如果一个在锁中执行了 sleep 之类的换出行为，是不是会发生调度性死锁。

**Reply by idealvin** [reply]
这是使用的问题，正常情况不会在释放锁前就sleep

**Reply by shapled** [reply]
封装之后很可能搞不清某个函数调用的里面是否有换入换出。

**Reply by idealvin** [reply]
> 封装之后很可能搞不清某个函数调用的里面是否有换入换出。

换入换出不会死锁，用户不需要关心这个

**Reply by shapled** [reply]
> 换入换出不会死锁，用户不需要关心这个

同 scheduler 内：
协程1 mutex_lock -> 协程1 换出 -> 协程 2 换入 -> 协程2 mutex_lock

后续即便协程1可以继续执行也无法切换回去了，这个行为是和线程下不一致的。

**Reply by idealvin** [reply]
不会出现这种情况，协程1没有释放锁，协程2不会获取到锁

**Reply by shapled** [reply]
就是取不到锁就把 scheduler 阻塞住了，协程 1 也再没有运行的机会了

**Reply by idealvin** [reply]
> 就是取不到锁就把 scheduler 阻塞住了，协程 1 也再没有运行的机会了

不会出现这种情况，pthread_mutex 那个在切换前已经释放了

**Comment by idealvin** [comment, marked as answer]
pthread_mutex_lock 并不是必须的，不会造成调度线程死锁
